servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DefaultDnsClient.java

```diff
@@ -424,9 +432,23 @@ protected AbstractDnsSubscription newSubscription(
             return new AbstractDnsSubscription(subscriber) {
                 @Override
                 protected Future<DnsAnswer<HostAndPort>> doDnsQuery(final boolean scheduledQuery) {
-                    Promise<DnsAnswer<HostAndPort>> promise = nettyIoExecutor.eventLoopGroup().next().newPromise();
-                    resolver.resolveAll(new DefaultDnsQuestion(name, SRV))
-                            .addListener((Future<? super List<DnsRecord>> completedFuture) -> {
+                    final EventLoop eventLoop = nettyIoExecutor.eventLoopGroup().next();
+                    final Promise<DnsAnswer<HostAndPort>> promise = eventLoop.newPromise();
+                    final Future<List<DnsRecord>> resolveFuture =
+                            resolver.resolveAll(new DefaultDnsQuestion(name, SRV));
+                    final Future<?> timeoutFuture = resolutionTimeoutMillis == 0L ? null : eventLoop.schedule(() -> {
+                        if (promise.isDone()) {
+                            return;
+                        }
+                        if (promise.tryFailure(DnsNameResolverTimeoutException.newInstance(name,
+                                resolutionTimeoutMillis, SRV.toString(), SrvRecordPublisher.class, "doDnsQuery"))) {
+                            resolveFuture.cancel(true);
+                        }
```

Comment:
ffti: it may not even change the bytecode but this could be a bit cleaner as
```
if (!promise.isDone() && promise.tryFailure(... )) {
  ...
}
```
Same below.